examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,612 @@
+# use the :class:`~sklearn.utils.metadata_routing.RequestType` values.
+
+est = (
+    ExampleClassifier()
+    .fit_requests(sample_weight=RequestType.UNREQUESTED)
+    .predict_requests(groups=RequestType.REQUESTED)
+    .score_requests(sample_weight=RequestType.UNREQUESTED)
+)
+print_routing(est)
+
+# %%
+# Please note that as long as the above estimator is not used in another
+# meta-estimator, the user does not need to set any requests for the metadata
+# and the set values are ignored, since a consumer does not validate or route
+# given metadata. A simple usage of the above estimator would work as expected.
+
+est = ExampleClassifier()
+est.fit(X, y, sample_weight=my_weights)
+est.predict(X[:3, :], groups=my_groups)
+
+# %%
+# Now let's have a meta-estimator, which doesn't do much other than routing the
+# metadata.
+
+
+class MetaClassifier(MetaEstimatorMixin, ClassifierMixin, BaseEstimator):
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def get_metadata_routing(self):
+        # This method defines the routing for this meta-estimator.
+        # In order to do so, a `MetadataRouter` instance is created, and the
+        # right routing is added to it. More explanation follows.
+        router = MetadataRouter(owner=self.__class__.__name__).add(
+            estimator=self.estimator, method_mapping="one-to-one"
+        )
+        return router
+
+    def fit(self, X, y, **fit_params):
+        # meta-estimators are responsible for validating the given metadata
+        # `get_router_for_object` is a safe way to construct a
+        # `MetadataRouter` from the given object.
+        request_router = get_router_for_object(self)
+        request_router.validate_metadata(params=fit_params, method="fit")
+        # we can use provided utility methods to map the given metadata to what
+        # is required by the underlying estimator. Here `method` refers to the
+        # parent's method, i.e. `fit` in this example.
+        routed_params = request_router.route_params(params=fit_params, caller="fit")
+
+        # the output has a key for each object's method which is used here,
+        # i.e. parent's `fit` method, containing the metadata which should be
+        # routed to them, based on the information provided in
+        # ``get_metadata_routing``.
+        self.estimator_ = clone(self.estimator).fit(X, y, **routed_params.estimator.fit)
+        self.classes_ = self.estimator_.classes_
+        return self
+
+    def predict(self, X, **predict_params):
+        check_is_fitted(self)
+        # same as in `fit`, we validate the given metadata
+        request_router = get_router_for_object(self)
+        request_router.validate_metadata(params=predict_params, method="predict")
+        # and then prepare the input to the underlying `predict` method.
+        params = request_router.route_params(params=predict_params, caller="predict")
```

Comment:
any reason why the validation and the getter of params are dissociated? can we make the validation part of route_params?

Reply:
people should use `process_routing` which does all of that anyway.